### PR TITLE
Implement Quarto chunk options UI

### PR DESCRIPTION
**Pull request by gtritchie**
### Intent

Addresses #11745

### Approach

Chunk / cell options specified using YAML syntax may now be edited via the gear icon / popup UI in both Quarto and R Markdown documents.

editing-ui

The UI will rewrite options back to the location they were found (e.g. first line or in the YAML block), but new options will go to the YAML location by default for Quarto, and the first line for R Markdown (preserving existing behaviour).

If an option exists both in the first line and in the YAML, then only the YAML one will be written back.

Options are sorted based on existing logic, only tweak I made was to have YAML "label:" be first, if present.

Option names will have periods as separators when in the first line, and dashes in the YAML (fig.cap vs fig-cap). Booleans are uppercase (TRUE/FALSE) in the R code (first line) and lowercase (true/false) in the YAML.

The current implementation does not support the `!expr` technique for including R code. If this or any other technique that the YAML parser doesn't understand are detected when clicking the gear icon, a message will be shown saying that the UI doesn't support the options and to edit manually.

In the future, we can look at customizing the schema used by the YAML parser to support the custom !expr tag (there are examples in the Quarto codebase), but I wanted to get this in before the Sun explodes.

### Automated Tests

There are some, I plan to author more.

### QA Notes

Verify you can edit chunk options whether they are in the first line, or in the YAML lines. Test in both Source mode and Visual mode, in both Quarto and R Markdown (R Markdown has always supported this syntax, not just Quarto).

If you put in a code expression like this (valid for Quarto) the UI should not appear and instead you'll get the warning dialog.

```YAML
#| fig-cap: !expr 'paste("Air", "Quality")' 
```
error-dialog

Any other text in the `#|` lines that isn't parsable as YAML will also trigger this.

### Documentation

None

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/tests/automation/testthat/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*.save
```
Comment (gtritchie) [inline src/cpp/tests/automation/testthat/.gitignore:1]: I often rename some of the test files to *.R.save to run a subset of tests, .git-ignoring them somewhat cuts down the clutter in the VSCode git UI.

Comment (kevinushey) [reply src/cpp/tests/automation/testthat/.gitignore:1]: Sounds good!

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/display/DefaultChunkOptionsPopupPanel.java [hunk 2]
```diff
@@ -224,14 +308,152 @@ else if (modeId == "mode/r")
...
+    */
+   private List<String> getYamlOptionLines(int startLine)
+   {
+      List<String> optionLines = new ArrayList<>();
+      int currentLine = startLine;
+      String line;
+      while ((line = display_.getLine(currentLine)) != null)
+      {
+         // a line with no characters after the starting #| isn't valid and ends detection
+         if (line.startsWith("#| ") && line.trim().length() > 2)
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/display/DefaultChunkOptionsPopupPanel.java:340]: Ultra-nit: if trailing whitespace is included, then content following that will be parsed by knitr. For example:

`````r
```{r}
#| example: true
#|           
#| example2: false
print(knitr::opts_current$get("example"))
print(knitr::opts_current$get("example2"))
```
`````

In this case, both `example` and `example2` will be set. Hopefully people aren't actually relying on this in the wild though; I imagine this is kind of "unspecified" behavior.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/display/DefaultChunkOptionsPopupPanel.java [hunk 3]
```diff
@@ -224,14 +308,152 @@ else if (modeId == "mode/r")
...
+    * @param startLine first potential option line
+    * @return number of lines found
+    */
+   private int countYamlOptionLines(int startLine)
+   {
+      int currentLine = startLine;
+      String line;
+      while ((line = display_.getLine(currentLine)) != null)
+      {
+         if (line.startsWith("#|"))
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/display/DefaultChunkOptionsPopupPanel.java:366]: It looks like the behavior here might differ between Quarto and R Markdown. In particular, R Markdown requires some whitespace following `#|`; e.g.

`````r
```{r}
#|example: true
print(knitr::opts_current$get("example"))
```
`````

When rendered, this will print "NULL" rather than TRUE. I'm not sure if this is true for Quarto as well, though.

Again, not sure if this is actually worth worrying about in practice, though!

**Review by kevinushey — APPROVED** [review]
Overall LGTM! I think some extra testing is warranted post-merge just to double check if there are any edge cases we need to be worried about (e.g. R Markdown vs. Quarto quirks in comment parsing).
